CRAYSAT-1549: Document SAT inclusion in CSM (#2366)
Cray-HPE/docs-csm: operations/sat/sat_in_csm.md

# System Admin Toolkit (SAT) in CSM

The System Admin Toolkit (SAT) is a command-line interface that can assist administrators with common tasks, such as
troubleshooting and querying information about the HPE Cray EX System, system boot and shutdown, and replacing hardware
components. In CSM 1.3 and newer, the `sat` command is available on the Kubernetes NCNs without installing the SAT
product stream.

> ## SAT Product Stream Components

Reviewer: Currently in the docs-csm repository we follow the HPE style guide, which (in the latest version we've seen) specifies that headings (other than the main page title) should use sentence casing (that is, the way you would capitalize a normal sentence).

```suggestion
## SAT product stream components
```

Author: Fixed in 14bd2c494e86e31f7460bf914b6d6cc49475dae4

It is still possible to install SAT as a separate product stream. Installing the SAT product stream results in
additional supporting components being installed.

- Installing the SAT product stream creates an entry for SAT in the `cray-product-catalog` Kubernetes ConfigMap.
  Without installing the SAT product stream, there will be no entry for this version of SAT in the output of
  `sat showrev`.

- The `sat-install-utility` container image is only available with the full SAT product stream. This container image
  provides uninstall and activate functionality when used with the `prodmgr` command. It is still possible to uninstall
  previous versions of the SAT product stream.

- The `sat-config-management` git repository in Gitea (VCS) and thus the SAT layer of NCN CFS configuration is

>   only available with the full SAT product stream. See [SAT CFS Configuration](#sat-cfs-configuration-details) below.

Reviewer: ```suggestion
  only available with the full SAT product stream. See [SAT CFS configuration details](#sat-cfs-configuration-details) below.
```

Author: Fixed in 14bd2c494e86e31f7460bf914b6d6cc49475dae4

- The `docs-sat` RPM package is only available with the full SAT product stream. SAT documentation can be found

>   at the links below (see: [SAT Documentation](#sat-documentation)).

Reviewer: ```suggestion
  at the links below (see: [SAT documentation](#sat-documentation)).
```

Author: Fixed in 14bd2c494e86e31f7460bf914b6d6cc49475dae4

> ### SAT CFS Configuration Details

Reviewer: ```suggestion
### SAT CFS configuration details
```

Author: Fixed in 14bd2c494e86e31f7460bf914b6d6cc49475dae4

> 

Reviewer 2: I recommend adding the heading "CFS configuration" back.

If we create a heading level, we want to make sure there is more than one. That way it doesn't oddly stick out on its own when there might be a better way to present the information. In your draft of this change for the SAT docs, the "CFS configuration" heading was on its own, so it was better to remove it. Here, though, we also have a second-level heading (##) of "SAT documentation," so I recommend adding "CFS configuration" back as a heading. That way we have two second-level headings, and "SAT documentation" isn't on its own.

That way there are two second-level headings for this section (not a single heading like in the SAT docs where we didn't have to link to the SAT documentation).

Author: I feel like in both cases it's supposed to be a child of "SAT product stream components" though, in that the SAT CFS configuration is one of the product stream components not available unless you install the full product stream. For that reason I originally had it as a 3rd-level heading in both locations - though as you pointed out we want to avoid having a new level of heading if there is only one.

> That way we have two second-level headings, and "SAT documentation" isn't on its own.

Right now we have two level-2 headings on this page, "SAT product stream components" and "SAT Documentation", right?

Reviewer 2: Man, my eyes are playing tricks on me. I agree with you, and you can disregard this comment :)

When the SAT product stream has not been installed, there will be no configuration content for SAT in VCS. Therefore,
CFS configurations that apply to NCNs (for example, `ncn-personalization`) should not include a SAT layer.

This configuration layer modifies the permissions of files left over from prior installs of SAT, so that the Keycloak
username used to authenticate to the API gateway cannot be read by users other than `root`. Specifically, it does the
following:

- Modify the `sat.toml` configuration file which contains the username so that it is only readable by `root`.

- Modify the `/root/.config/sat/tokens` directory so that the directory is only readable by `root`, because the
  names of the files within the `tokens` directory contain the username.

Regardless of this configuration being applied, the contents of the tokens are never readable by other users, nor are
any passwords. These permission changes only apply to files created by previous installs of SAT; in the current version
of SAT all files and directories are created with the appropriate permissions.

> ## SAT Documentation

Reviewer: ```suggestion
## SAT documentation
```

Author: Fixed in 14bd2c494e86e31f7460bf914b6d6cc49475dae4

> For full SAT documentation, please see the [SAT Documentation](https://cray-hpe.github.io/docs-sat/en-24).

Reviewer: ```suggestion
For full SAT documentation, see the [SAT Documentation](https://cray-hpe.github.io/docs-sat/en-24).
```

> For full SAT documentation, see the [SAT Documentation](https://cray-hpe.github.io/docs-sat/).

Reviewer 2: ```suggestion
For full SAT documentation, refer to the [*HPE Cray EX System Admin Toolkit (SAT) Guide*](https://cray-hpe.github.io/docs-sat/).
```

> If SAT has not been installed before, some initial configuration is required, for example to authenticate to the API

Reviewer: ```suggestion
If SAT has not been installed before, then some initial configuration is required; for example to authenticate to the API
```

Author: Fixed in 14bd2c494e86e31f7460bf914b6d6cc49475dae4

gateway with `sat auth`.

- [SAT Authentication](https://cray-hpe.github.io/docs-sat/en-24/install/#sat-authentication)
- [Generate SAT S3 Credentials](https://cray-hpe.github.io/docs-sat/en-24/install/#generate-sat-s3-credentials)
- [Set System Revision Information](https://cray-hpe.github.io/docs-sat/en-24/install/#set-system-revision-information)

> If the full SAT product stream is not being installed, it may be wise to uninstall old versions of the SAT product

Reviewer: ```suggestion
If the full SAT product stream is not being installed, then it may be wise to uninstall old versions of the SAT product
```

Author: Fixed in 14bd2c494e86e31f7460bf914b6d6cc49475dae4

> gateway with `sat auth`. To complete the initial configuration of SAT, complete the following processes:
> 
> - [SAT Authentication](https://cray-hpe.github.io/docs-sat/en-24/install/#sat-authentication)
> - [Generate SAT S3 Credentials](https://cray-hpe.github.io/docs-sat/en-24/install/#generate-sat-s3-credentials)
> - [Set System Revision Information](https://cray-hpe.github.io/docs-sat/en-24/install/#set-system-revision-information)

Reviewer 2: ```suggestion
gateway with `sat auth`. To complete the initial configuration of SAT, refer to the following processes in the SAT documentation:

- SAT Authentication
- Generate SAT S3 Credentials
- Set System Revision Information
```

> If SAT has not been installed before, some initial configuration is required, for example, to authenticate to the API
> gateway with `sat auth`. To complete the initial configuration of SAT, complete the following processes:
> 
> - [SAT Authentication](https://cray-hpe.github.io/docs-sat/en-24/install/#sat-authentication)
> - [Generate SAT S3 Credentials](https://cray-hpe.github.io/docs-sat/en-24/install/#generate-sat-s3-credentials)
> - [Set System Revision Information](https://cray-hpe.github.io/docs-sat/en-24/install/#set-system-revision-information)

Reviewer 2: ```suggestion
If SAT has not been installed before, some initial configuration is required (for example, to authenticate to the API
gateway with `sat auth`). To complete the initial configuration of SAT, refer to the following processes in the SAT documentation:

- **SAT Authentication**
- **Generate SAT S3 Credentials**
- **Set System Revision Information**
```

> If the full SAT product stream is not being installed, then it is recommended that you uninstall old versions of the

Reviewer 2: ```suggestion
If the full SAT product stream is not being installed, it is recommended that you uninstall old versions of the
```

> SAT product stream to avoid confusion in the output of `sat showrev`.
> 
> See [Uninstall: Removing a Version of SAT](https://cray-hpe.github.io/docs-sat/en-24/install/#uninstall-removing-a-version-of-sat)
> for more information.

Reviewer 2: ```suggestion
SAT product stream to avoid confusion in the output of `sat showrev`. For more information, refer to **Uninstall: Removing a Version of SAT** in the SAT documentation.
```

# System Admin Toolkit (SAT) in CSM

The System Admin Toolkit (SAT) is a command-line interface that can assist administrators with common tasks, such as
troubleshooting and querying information about the HPE Cray EX System, system boot and shutdown, and replacing hardware
components. In CSM 1.3 and newer, the `sat` command is available on the Kubernetes NCNs without installing the SAT
product stream.

## SAT product stream components

It is still possible to install SAT as a separate product stream. Installing the SAT product stream results in

> additional supporting components being installed.

Author: Addressed in 8bcd472a677be33563916de87f0748c6c5647235

- Installing the SAT product stream creates an entry for SAT in the `cray-product-catalog` Kubernetes ConfigMap.
  Without installing the SAT product stream, there will be no entry for this version of SAT in the output of

>   `sat showrev`.

Author: Addressed in 8bcd472a677be33563916de87f0748c6c5647235

- The `sat-install-utility` container image is only available with the full SAT product stream. This container image
  provides uninstall and activate functionality when used with the `prodmgr` command. It is still possible to uninstall

>   previous versions of the SAT product stream.

Author: Addressed in 8bcd472a677be33563916de87f0748c6c5647235 and 6820112535df9774aa5cde61c17e1f9746638f18

- The `sat-config-management` git repository in Gitea (VCS) and thus the SAT layer of NCN CFS configuration is
  only available with the full SAT product stream. See [SAT CFS configuration details](#sat-cfs-configuration-details)
  below.

- The `docs-sat` RPM package is only available with the full SAT product stream. SAT documentation can be found

>   at the links below (see: [SAT documentation](#sat-documentation)).

Author: Addressed in 8bcd472a677be33563916de87f0748c6c5647235

> ### SAT CFS configuration details

Reviewer 2: I'd suggest making this a level two heading like the others under "SAT in CSM."
```suggestion
## SAT CFS configuration details
```

Author: Would it be OK if I remove the heading entirely? Then this section will be nearly identical to the one we have in docs-sat.

> When the SAT product stream has not been installed, there will be no configuration content for SAT in VCS. Therefore,

Author: Addressed in 8bcd472a677be33563916de87f0748c6c5647235 and 6820112535df9774aa5cde61c17e1f9746638f18

CFS configurations that apply to NCNs (for example, `ncn-personalization`) should not include a SAT layer.

This configuration layer modifies the permissions of files left over from prior installs of SAT, so that the Keycloak
username used to authenticate to the API gateway cannot be read by users other than `root`. Specifically, it does the
following:

- Modify the `sat.toml` configuration file which contains the username so that it is only readable by `root`.

- Modify the `/root/.config/sat/tokens` directory so that the directory is only readable by `root`, because the
  names of the files within the `tokens` directory contain the username.

Regardless of this configuration being applied, the contents of the tokens are never readable by other users, nor are
any passwords. These permission changes only apply to files created by previous installs of SAT; in the current version
of SAT all files and directories are created with the appropriate permissions.

## SAT documentation

> For full SAT documentation, see the [SAT Documentation](https://cray-hpe.github.io/docs-sat/en-24).

Reviewer 3: That works for me too. Seems like a good idea.

> For full SAT documentation, see the [SAT Documentation](https://cray-hpe.github.io/docs-sat/en-24).

Author: It does, but unfortunately that does not apply to the remaining links, i.e. "https://cray-hpe.github.io/docs-sat/install/#sat-authentication" does not work.

> For full SAT documentation, see the [SAT Documentation](https://cray-hpe.github.io/docs-sat/en-24).

Reviewer 2: Yeah, I wasn't sure if it would since that URL is more specific. I'll try to talk to my team next week about the best way to link/reference between documentation. You can always merge this PR with the more specific links as they are, and I can update them later if we find a better way to maintain them.

> For full SAT documentation, see the [SAT Documentation](https://cray-hpe.github.io/docs-sat/en-24).

Author: I think it makes sense to strip off the version-specific component where we can, at least for this first one we can do that.

If SAT has not been installed before, then some initial configuration is required; for example to authenticate to the

> API gateway with `sat auth`.

Author: There's a very small inaccuracy in the suggestion here in that authenticating with `sat auth` is actually just the first process in the list. The reader should do all 3 but I just gave the first as an example. This would probably be more accurate:

```
If SAT has not been installed before, some initial configuration is required, for example, to authenticate to the API
gateway with `sat auth`. To complete the initial configuration of SAT, complete the following processes:
```

It probably could be reworded even further.

> API gateway with `sat auth`.

Author: I implemented the above in 8bcd472a677be33563916de87f0748c6c5647235 - but would be open to refining that further.

> API gateway with `sat auth`.

Reviewer 2: I tweaked your wording a little, but I like your new approach.

- [SAT Authentication](https://cray-hpe.github.io/docs-sat/en-24/install/#sat-authentication)
- [Generate SAT S3 Credentials](https://cray-hpe.github.io/docs-sat/en-24/install/#generate-sat-s3-credentials)

> - [Set System Revision Information](https://cray-hpe.github.io/docs-sat/en-24/install/#set-system-revision-information)

Reviewer 3: Yeah, it would be nice if we didn't have to keep these up to date.

> - [Set System Revision Information](https://cray-hpe.github.io/docs-sat/en-24/install/#set-system-revision-information)

Author: Agreed - going forward we could probably implement some kind of "https://cray-hpe.github.io/docs-sat/en-latest/" type link so that the version is not hard-coded in the URL. That would require some code changes to the scripts that render the HTML docs.

> - [Set System Revision Information](https://cray-hpe.github.io/docs-sat/en-24/install/#set-system-revision-information)

Reviewer 2: Eli, that sounds like an idea worth exploring! Until we're able to do this, I think we'll need to remove these links since we don't want the overhead of maintaining them.

We can include the section names for reference, and users should be able to search for them within the SAT documentation. It helps that we provide a link to the general SAT documentation earlier in this section. (We'll just want to avoid renaming sections of our SAT documentation unless absolutely necessary.) I confirmed that this is currently my team's recommended approach for references between EX documents.

> If the full SAT product stream is not being installed, then it may be wise to uninstall old versions of the SAT product

Author: Addressed in 8bcd472a677be33563916de87f0748c6c5647235

stream to avoid confusion in the output of `sat showrev`.

> - [Uninstall: Removing a Version of SAT](https://cray-hpe.github.io/docs-sat/en-24/install/#uninstall-removing-a-version-of-sat)

Author: Addressed in 8bcd472a677be33563916de87f0748c6c5647235
